### ISSUE TITLE
Can not response catchall route result with ssl

### ISSUE THREAD

**Issue by whyjp** [CrowCpp/Crow#833]
when i not include ssl cert.
i have do correctly resposnse with catch_all route. 
in dynamic route on codes

but i put ssl and call with https i has not response to client 
error : Error: socket hang up
and on web brouser : ERR_RESPONSE_HEADERS_TRUNCATED

when i debug codes

in async_write rise error with error codes message worng filehandle

**Comment by gittiver** [comment]
If there is no SSL cert, how should a request or response be encrypted? therefore there can't be any answer on catch_all.

**Comment by whyjp** [comment]
@gittiver 

Sorry, the question was misdirected.

In other words
Whether ssl is set or not
catch_all in progress
Disconnecting from async_read

self->adaptor_.close(); //http_connection.h:504

Doesn't this seem a bit wrong?

My intention is that if the path is intended in catch_all, even if it is not predefined,
I want to send the response normally as 200.

When setting up SSL, an error occurs when trying to write to the socket disconnected in this code.

**Comment by whyjp** [comment]
when i skip this code condition in debug 
```
if (error_while_reading)
                  {
                      self->cancel_deadline_timer();
                      self->parser_.done();
                      self->adaptor_.shutdown_read();
                      self->adaptor_.close();
                      CROW_LOG_DEBUG << self << " from read(1) with description: \"" << http_errno_description(static_cast<http_errno>(self->parser_.http_errno)) << '\"';
                  }
```
catch_all is success response to client

**Comment by whyjp** [comment]
@gittiver 

my explanation was wrong

1. I tried to implement a dynamic variable path such as path/value1/path2/value2 that was not predefined through catch_all.
Even if this was possible, until SSL was used...

When SSL is not used (no certificate is included)
It worked well this time, as intended.

However, after inserting the certificate to use SSL,
All general paths operate normally, but
through catch_all
The code that tried to respond normally by changing the 400 response code to 200 no longer worked properly.

In async_write, an error value of false appeared in ec.
At this time, the error said that the wrong file handle was accessed.

**Comment by gittiver** [comment]
well, with ssl there is an error on read, as there is no cert, therefore no handshake, therefore no decryption key? 
So closing socket is Ok?

**Comment by whyjp** [comment]
No i put the cert by self signed 
so it works good regular paths. like `loaclhost:10000/path1`
but does not works in catch_all routes with ssl

**Comment by gittiver** [comment]
so you have an ssl cert but it is self signed and only the catch_all route is not working, all others are working? correct?

**Comment by whyjp** [comment]
That's right, all general paths are working well.

When implementing catch_all to redirect to the correct path
Shouldn't we just change the code/body values ​​of response obj and then call res.end()?

Current code is
came in as catch_all
(with crow's 400 res code)
If the path is as intended, store and process the value.
res.code = 200 | res.body = "res text"
re.end()
I finished the code and felt that it worked well when not using SSL.

**Comment by whyjp** [comment]
compile time checkable variable handler
is very useful but

I am not currently implementing the webserver directly, but I was trying to implement a more dynamic variable path.
catch_all was a valuable feature not found in other web frameworks.

However, after adding SSL, it did not work as desired.
Looking at the debug log, an error occurred in each read/write.

**Comment by gittiver** [comment]
sounds like a bug. rewrote the title too.

**Comment by whyjp** [comment]
i try adding catchall handler 
like 
```
auto catchall = [](const crow::request& req, crow::response& res) {
        if (req.method == crow::HTTPMethod::GET) {
            std::ostringstream os;
            os << "req.url: " << req.url << "\n\n";
            os << "Params: " << req.url_params << "\n\n";
            printf("%s", os.str().c_str());
            res = crow::response(200, "application/json", "res force OK");
            return;
  }
};
```

with ssl cert 
it cant response 

thx your answer
i trying too find a bug codes in repo

**Comment by whyjp** [comment]
i found key points this bugs..

when use ssl cert.


self->adaptor_.close();
in do_read()

before write ends

**Comment by whyjp** [comment]
in this codes 


```
void handle_url()
        {
            routing_handle_result_ = handler_->handle_initial(req_, res);
            // if no route is found for the request method, return the response without parsing or processing anything further.
            if (!routing_handle_result_->rule_index)
            {
                parser_.done();
                need_to_call_after_handlers_ = true;
                complete_request();
            }
        }
```
handle_initial
If catchall_route is encountered internally, some additional processing may be required.
Even if rule_index is 0, you should not close the socket using the code below.
Especially in the case of https

However, in the case of ssl + catchall, entering the if condition of the above code
The socket is closed before the write is completed.
A bug occurred only when using SSL.

Simply put
Only when catchall is inside get_error
found.rule_index = -1;
I confirmed that the problem was solved by temporarily adding the code, but

A more obvious code modification is

I think it would be better to have someone else who is already familiar with Crow analyze and solve the problem.
I will not reflect the code

**Comment by whyjp** [comment]
Because the actual cause of the bug itself is
After reading ssl + request
This is because the socket is closed because it is not handled statically.
Normally, if you want to respond even with 400, you should not close the socket like this.

This bug is not limited to catch_all after all
Confirmed with ssl and handler case

I hope you change the title
@gittiver

**Comment by paulharris** [comment]
I think this bug is related to all the other catchall bugs, eg #669 and #758 

I see similar behaviour when running crowcpp as HTTP, behind a reverse proxy.
When hitting a missing route directly on the http port, I get a 404.

When I hit that same route via the https reverse proxy, I get a 502-bad-gateway from the proxy server!

This is the debug log when I create a route for "/<path>" as my catch-all,
```
(2024-11-13 07:28:39) [DEBUG   ] 0x55555632dd20 {2} queue length: 1
(2024-11-13 07:28:39) [DEBUG   ] 0x7fff800010d0 timer cancelled: 0x7ffff33fd648 0
(2024-11-13 07:28:39) [DEBUG   ] task_timer task cancelled: 0x7ffff33fd648 0
(2024-11-13 07:28:39) [DEBUG   ] task_timer scheduled: 0x7ffff33fd648 1
(2024-11-13 07:28:39) [DEBUG   ] 0x7fff800010d0 timer added: 0x7ffff33fd648 1
(2024-11-13 07:28:39) [DEBUG   ] 0x7fff800010d0 timer cancelled: 0x7ffff33fd648 1
(2024-11-13 07:28:39) [DEBUG   ] task_timer task cancelled: 0x7ffff33fd648 1
(2024-11-13 07:28:39) [DEBUG   ] Matched rule '/<path>' 1 / 2
(2024-11-13 07:28:39) [DEBUG   ] 0x7fff800010d0 timer cancelled: 0x7ffff33fd648 1
(2024-11-13 07:28:39) [DEBUG   ] task_timer task cancelled: 0x7ffff33fd648 1
(2024-11-13 07:28:39) [DEBUG   ] task_timer scheduled: 0x7ffff33fd648 2
(2024-11-13 07:28:39) [DEBUG   ] 0x7fff800010d0 timer added: 0x7ffff33fd648 2
(2024-11-13 07:28:39) [DEBUG   ] 0x7fff800010d0 timer cancelled: 0x7ffff33fd648 2
(2024-11-13 07:28:39) [DEBUG   ] task_timer task cancelled: 0x7ffff33fd648 2
(2024-11-13 07:28:39) [DEBUG   ] Matched rule '/login' 1 / 2
(2024-11-13 07:28:39) [DEBUG   ] 0x7fff800010d0 timer cancelled: 0x7ffff33fd648 2
(2024-11-13 07:28:39) [DEBUG   ] task_timer task cancelled: 0x7ffff33fd648 2
(2024-11-13 07:28:39) [DEBUG   ] task_timer scheduled: 0x7ffff33fd648 3
(2024-11-13 07:28:39) [DEBUG   ] 0x7fff800010d0 timer added: 0x7ffff33fd648 3
(2024-11-13 07:28:40) [DEBUG   ] 0x7fff800010d0 timer cancelled: 0x7ffff33fd648 3
(2024-11-13 07:28:40) [DEBUG   ] task_timer task cancelled: 0x7ffff33fd648 3
(2024-11-13 07:28:40) [DEBUG   ] 0x7fff800010d0 from read(1) with description: "success"
```

And this is the debug log when I let the default (or custom) catchall respond:
```
(2024-11-13 07:30:38) [DEBUG   ] 0x5555562fc5d0 {0} queue length: 1
(2024-11-13 07:30:47) [DEBUG   ] 0x55555632cb50 {1} queue length: 1
(2024-11-13 07:30:47) [DEBUG   ] 0x555556334270 timer cancelled: 0x7ffff3bfe648 0
(2024-11-13 07:30:47) [DEBUG   ] task_timer task cancelled: 0x7ffff3bfe648 0
(2024-11-13 07:30:47) [DEBUG   ] task_timer scheduled: 0x7ffff3bfe648 1
(2024-11-13 07:30:47) [DEBUG   ] 0x555556334270 timer added: 0x7ffff3bfe648 1
(2024-11-13 07:30:47) [DEBUG   ] Cannot match rules /something. 
(2024-11-13 07:30:47) [DEBUG   ] 0x555556334270 timer cancelled: 0x7ffff3bfe648 1
(2024-11-13 07:30:47) [DEBUG   ] task_timer task cancelled: 0x7ffff3bfe648 1
(2024-11-13 07:30:47) [DEBUG   ] 0x555556334270 from read(1) with description: "stream ended at an unexpected time"
(2024-11-13 07:30:47) [DEBUG   ] 0x555556334270 from write(2)
```

Note the "stream ended at an unexpected time", doing a quick trace looks like it loops back on itself.
I think it looks a bit like like the pattern in this bug report (ie continuing to try and read when it shouldn't):
https://github.com/CrowCpp/Crow/pull/799#issuecomment-2417316854

**Comment by paulharris** [comment]
Also note that a manual "catch all" `/<path>` route does not work perfectly,
as I have to specify all the methods that route needs to capture,
so eg if I do `.methods(crow::HTTPMethod::Get)` then a POST to a missing route will generate a 502 from the reverse proxy.
And if I specify ::Post, then the Get will generate 502.
I can specify Get, Post, but I'm not going to specify Put and all the other ones...
So still want a catchall that works...

And I'd want it to go through my middleware and complain that the call was not authenticated, so bad actors can't probe for valid URLs without authentication.

**Comment by gittiver** [comment]
fixed by linked PR
